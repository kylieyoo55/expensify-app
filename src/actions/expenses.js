import { v4 as uuidv4 } from 'uuid';


//add_expense
export const addExpense=({
    description='',
    note='',
    amount=0,
    createAt=0
}={})=>({
type:'add_expense',

expense:{
    id:uuidv4(),
    description,
    note,
    amount,
    createAt
}
})

// remove_expense

export const removeExpense=({id})=>(
    {
        type:'remove_expense',
        id
    }
)

//edit_expense
export const  editExpense=(id,updates)=>(
    {
        type:'edit_expense',
        id,
        updates
    }
)