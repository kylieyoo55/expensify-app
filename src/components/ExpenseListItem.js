import React from 'react';

import { Link } from 'react-router-dom';

//export a staless function component
//description,amount,createAt showing on Screen




const ExpenseListItem = ({id, description, amount, createAt})=>

    (
<div  >
<Link to={`/edit/${id}`}>
<h2>Expense List Item</h2>
</Link>
<p>{description}</p>
<p>Amount: {amount}- Created: {createAt}</p>

</div>
)


export default ExpenseListItem;