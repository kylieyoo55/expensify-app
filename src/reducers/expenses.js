

const expensesReducerDefault=[]

const expensesReducer=(state=expensesReducerDefault,action)=>{
    switch(action.type){
        case'add_expense':
            return [
             ...state,
            action.expense];

        case 'remove_expense':
            return state.filter(({id})=> id !== action.id);
        
        case 'edit_expense':
            return state.map((exp)=>{
               if( exp.id === action.id){
                   return{
                       ...exp,
                       ...action.updates
                   }
               }else{return exp}
            })

        default: 
            return state;    }
};

export default expensesReducer;