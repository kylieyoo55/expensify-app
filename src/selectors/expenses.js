import moment from 'moment';

const getVisibleExpenses=(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((exp)=>{
        const createAtMoment=moment(exp.createAt);
        const startDateMatch = startDate? startDate.isSameOrBefore(createAtMoment,'day') :true;
        const endDateMatch= endDate?endDate.isSameOrAfter(createAtMoment,'day'):true;
        const textMatch=exp.description.toLowerCase().includes(text.toLowerCase());
        // typeof startDate !== 'number' || exp.createAt >= startDate;
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
if(sortBy === 'date'){
    return a.createAt < b.createAt?1:-1;
}
if(sortBy ==='amount'){
    return a.amount < b.amount? 1:-1;
}
return 0;
    });
}

export default getVisibleExpenses;