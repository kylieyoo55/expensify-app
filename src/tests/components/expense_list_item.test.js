import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixture/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

//create test file
test('should render expensive list item with expenses',()=>{
    const wrapper=shallow(<ExpenseListItem key={expenses[0].id} {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
    })