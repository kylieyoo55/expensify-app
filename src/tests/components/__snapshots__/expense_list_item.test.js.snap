// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expensive list item with expenses 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h2>
      Expense List Item
    </h2>
  </Link>
  <p>
    Gum
  </p>
  <p>
    Amount: 
    190
    - Created: 
    0
  </p>
</div>
`;
